src/SlidingWindow.cpp

```diff
@@ -119,6 +119,94 @@ bool find_produce(const Stmt &s, const string &func) {
     return finder.found;
 }
 
+// This mutator rewrites calls and provides to a particular
+// func:
+// - Calls and Provides are shifted to be relative to the min.
+// - Provides additionally are rewritten to load values from the
+//   previous iteration of the loop if they were computed in the
+//   last iteration.
+class RollFunc : public IRMutator {
+    const Function &func;
+    int dim;
+    const string &loop_var;
+    const Interval &old_bounds;
+    const Interval &new_bounds;
+
+    // It helps simplify the shifted calls/provides to rebase the
+    // loops that are subtracted from to have a min of 0.
+    set<string> loops_to_rebase;
+    bool in_produce = false;
+
+    using IRMutator::visit;
+
+    Stmt visit(const ProducerConsumer *op) override {
+        bool produce_func = op->name == func.name() && op->is_producer;
+        ScopedValue<bool> old_in_produce(in_produce, in_produce || produce_func);
+        return IRMutator::visit(op);
+    }
+
+    Stmt visit(const Provide *op) override {
+        if (!(in_produce && op->name == func.name())) {
+            return IRMutator::visit(op);
+        }
+        vector<Expr> values = op->values;
+        for (Expr &i : values) {
+            i = mutate(i);
+        }
+        vector<Expr> args = op->args;
+        for (Expr &i : args) {
+            i = mutate(i);
+        }
+        bool sliding_up = old_bounds.max.same_as(new_bounds.max);
+        Expr is_new = sliding_up ? new_bounds.min <= args[dim] : args[dim] <= new_bounds.max;
+        args[dim] -= old_bounds.min;
+        vector<Expr> old_args = args;
+        old_args[dim] = substitute(loop_var, Variable::make(Int(32), loop_var) - 1, old_args[dim]);
```

Comment:
Doesn't this need an equivalent of expand_expr? What if the dependence on the loop var is via a letstmt?